engine/classes/Elgg/Application/ServeFileHandler.php

```diff
@@ -51,7 +51,7 @@ public function getResponse(Request $request) {
 		$response = new Response();
 		$response->prepare($request);
 
-		$path = implode('/', $request->getUrlSegments());
+		$path = implode('/', $request->getUrlSegments(true));
```

Comment:
If file name has `&` it fails nevertheless - Symfony doesn't parse segments properly